source/utf8/unchecked.h

```diff
@@ -215,51 +223,214 @@ namespace utf8
+            using iterator_category = std::bidirectional_iterator_tag;
+            using value_type = uint32_t;
+            using difference_type = ptrdiff_t;
+            using pointer = uint32_t*;
+            using reference = uint32_t&;
+
+            bidirectional_iterator () {}
+            explicit bidirectional_iterator (const octet_iterator& octet_it): it(octet_it) {}
             // the default "big three" are OK
             octet_iterator base () const { return it; }
             uint32_t operator * () const
             {
                 octet_iterator temp = it;
                 return utf8::unchecked::next(temp);
             }
-            bool operator == (const iterator& rhs) const 
+            bool operator == (const bidirectional_iterator& rhs) const 
             { 
                 return (it == rhs.it);
             }
-            bool operator != (const iterator& rhs) const
+            bool operator != (const bidirectional_iterator& rhs) const
             {
                 return !(operator == (rhs));
             }
-            iterator& operator ++ () 
+            bidirectional_iterator& operator ++ () 
             {
                 ::std::advance(it, utf8::internal::sequence_length(it));
                 return *this;
             }
-            iterator operator ++ (int)
+            bidirectional_iterator operator ++ (int)
             {
-                iterator temp = *this;
-                ::std::advance(it, utf8::internal::sequence_length(it));
+                bidirectional_iterator temp = *this;
+                ++(*this);
                 return temp;
             }  
-            iterator& operator -- ()
+            bidirectional_iterator& operator -- ()
             {
                 utf8::unchecked::prior(it);
                 return *this;
             }
-            iterator operator -- (int)
+            bidirectional_iterator operator -- (int)
             {
-                iterator temp = *this;
-                utf8::unchecked::prior(it);
+                bidirectional_iterator temp = *this;
+                --(*this);
                 return temp;
             }
-          }; // class iterator
+          }; // class bidirectional_iterator
+
+        template <typename octet_iterator>
+        class input_iterator : public std::iterator<std::input_iterator_tag, uint32_t> {
+          private:
+          octet_iterator it;
+          uint32_t cp{};
+          void read() {
+              cp = utf8::unchecked::next_impl(it);
+          }
+    
+          public:
+          using iterator_category = std::input_iterator_tag;
+          using value_type = uint32_t;
+          using difference_type = ptrdiff_t;
+          using pointer = uint32_t*;
+          using reference = uint32_t&;
+
+          input_iterator () {}
+          explicit input_iterator (const octet_iterator& octet_it) : it(octet_it)
+          {
+              read();
```

Comment:
If this is constructed from the end-iterator you are dereferencing the end iterator which is UB AFAIK. Example where this might crash: Non-NULL terminated string allocated in an area just large enough. The end-iter points to invalid, undereferencable memory.

Reply:
You're right. The istream_iterator returns 0 in this case but this is UB. May be I should add an extra parameter in the constructor (bool ?) to enable/disable the read action. By default, the constructor reads the code point. In class utf8_sequence_stream, I can easily call the right constructor.
Have you an other suggestion ?

Reply:
I think this is impossible to solve without also passing the end iterator. You have the same problem in the `++`-op: You increment the underlying `it` and then call `read` w/o checking if you actually can do that. This is why I did not implement the unchecked iterator in my PR. So either one has to pass the end-iterator for input iterators or you have to read the value on dereference. Something like:

- Ctor sets `unread` flag
- `*operator` checks for flag, if set, read and cache and unset flag, else return cached value
- `++operator` checks for flag, if set read, either way set flag

But this probably makes the comparison impossible because after `++it`  "Any copies of the previous value of it are no longer required to be either dereferenceable or to be in the domain of ==. "

BTW: I'm not a big fan of having 6 classes for those. Why not make it a pair of the begin/end iterator with methods begin/end similar to what you had before? This allows usage in range-based for and is just the essentials.

Reply:
You're right, `++op` needs the end-iterator. My conclusion is : trying to manipulate an `unchecked::iterator` with an `input-iterator` must be a compile-time error (`static-assert` probably).
I will think about the 6 classes problem.